### ISSUE TITLE
IPP-DEV-09-001: Create a simple KNN algorithm implementation

### ISSUE THREAD

**Issue by labbenchstudios** [programming-in-python/ipp-exercise-tasks#84]
# Description
- Using instructions from the in-class lecture, create a module and class that implements the KNN algorithm.
- This module will need to implement the functionality listed below under Actions.

### Review the README
- Please see [README.md](https://github.com/programming-in-python/ipp-exercise-tasks) for further information on, and use of, this content.
- License for embedded documentation and source codes: [IPP-DOC-LIC](https://github.com/programming-in-python/ipp-exercise-tasks/issues/2)

### Estimated effort may vary greatly
- The estimated level of effort for this exercise shown in the 'Estimate' section below is a very rough approximation. The actual level of effort may vary greatly depending on your development and test environment, experience with the requisite technologies, and many other factors.

### Reminders on System Configuration:
- Make Sure Your System is Setup for Python and This Course
  - See [IPP-CFG-01-001](https://github.com/programming-in-python/ipp-exercise-tasks/issues/29)

- Make Sure PYTHONPATH is Set Correctly
  - Whether running Python tests within your IDE or from the command line, you must set the PYTHONPATH environment variable in every execution environment (e.g., every terminal you launch) when attempting to run any of your scripts and their tests or the IPP test app from the command line. The IPP source and test paths will be as follows:
    - {your IPP source code path}
    - {your IPP source code path}/tests
  - As a reminder, you can include the PYTHONPATH environment variable for your environment within the activation script for your virtual environment, which will set PYTHONPATH automatically whenever your virtual environment is launched

- See [IPP-DEV-01-001](https://github.com/programming-in-python/ipp-exercise-tasks/issues/40) for details.

### Step 1: In the `IPP_HOME/exercises/labmodule09` path, create a new module named `SimpleKnnAlgorithm.py` and add the following:
  - Declare the import statements
  - Declare the class to be named `SimpleKnnAlgorithm`.
- Here's a BRIEF example of the class definition with its important statements:

```python
import math

from typing import List, Tuple, Optional, Dict

class SimpleKnnAlgorithm:
```

### Step 2: Create the constructor
- Create the constructor
  - This will perform a few important tasks
    - Sets the number of neighbors to consider ('k')
    - Initializes the training data list and the feature names list
  - Here's an implementation approach for the constructor:

```python
    def __init__(self, k: int = 3):
        self.k = k
        self.trainingData = []
        self.featureNames = []
```

### Step 3: Create a setter for capturing training data
- This very simple method with a complex data structure as its parameter (`List[Tuple[List[float], float]]`).
  - Be sure to validate the incoming data - at least make sure it's not None!
- Here's one implementation approach:

```python
    def setTrainingData(self, data: List[Tuple[List[float], float]]):
        if not data:
            print("Warning: No training data provided")
            return
        
        self.trainingData = data
        print(f"KNN trained with {len(data)} samples")
```

### Step 4: Create the methods to calculate the distance between two points
- This will use both Euclidean distance (as the crow flies) and Manhattan distance (follow a grid pattern)
- Here's an example implementation for calculating Euclidean distance:

```python
    def calculateEuclideanDistance(self, point1: List[float], point2: List[float]) -> float:
        if len(point1) != len(point2):
            return float('inf')
        
        sumSquares = 0.0
        for i in range(len(point1)):
            diff = point1[i] - point2[i]
            sumSquares += diff * diff
        
        return math.sqrt(sumSquares)
```

- Here's an example implementation for calculating Manhattan distance:

```python
    def calculateManhattanDistance(self, point1: List[float], point2: List[float]) -> float:
        if len(point1) != len(point2):
            return float('inf')
        
        sumAbs = 0.0
        for i in range(len(point1)):
            sumAbs += abs(point1[i] - point2[i])
        
        return sumAbs
```

### Step 5: Create a method to find the nearest neighbor using either Euclidean or Manhattan distance:
- Here's an example implementation:

```python
    def findKNearest(self, testPoint: List[float], useEuclidean: bool = True) -> List[Tuple[float, float]]:
        if not self.trainingData:
            return []
        
        # Calculate distances to all training points
        distances = []
        for features, target in self.trainingData:
            if useEuclidean:
                dist = self.calculateEuclideanDistance(testPoint, features)
            else:
                dist = self.calculateManhattanDistance(testPoint, features)

            # Append a single tuple
            distances.append((dist, target))
        
        # Sort by distance and return k nearest
        distances.sort(key = lambda x: x[0])
        return distances[:self.k]
```

### Step 6: Create the regression and classification prediction methods:
- This step requires the implementation of `predictRegression()` and `predictClassification()`. The former predicts a continuous value using KNN regression, while the latter predicts a class label using KNN classifiation.
- Here's an example of both methods:
```python
    def predictRegression(self, testPoint: List[float], useEuclidean: bool = True) -> Optional[float]:
        neighbors = self.findKNearest(testPoint, useEuclidean)
        
        if not neighbors:
            return None
        
        # Calculate average of neighbor values
        total = sum(target for _, target in neighbors)
        prediction = total / len(neighbors)
        
        return prediction
    
    def predictClassification(self, testPoint: List[float], useEuclidean: bool = True) -> Optional[int]:
        neighbors = self.findKNearest(testPoint, useEuclidean)
        
        if not neighbors:
            return None
        
        # Count votes for each class
        votes = {}
        for _, target in neighbors:
            targetClass = int(target)
            votes[targetClass] = votes.get(targetClass, 0) + 1
        
        # Return class with most votes
        prediction = max(votes, key = votes.get)
        return prediction
```

### Step 7: Create a method to evaluate prediction error rates
- This is a very simple method that evaluates the 'actual' value and 'predicted' value:

```python
    def evaluatePrediction(self, actual: float, predicted: float) -> Dict[str, float]:
        error = actual - predicted
        absError = abs(error)
        
        # Calculate percentage error (avoid division by zero)
        if actual != 0:
            percentError = (absError / abs(actual)) * 100
        else:
            percentError = 0 if predicted == 0 else float('inf')
        
        return {
            'error': error,
            'absolute_error': absError,
            'squared_error': error * error,
            'percent_error': percentError
        }
```

## Estimate
- Medium

### Create the test module `test_SimpleKnnAlgorithm.py` within the `tests/labmodule09` path, and the class `SimpleKnnAlgorithmTest(unittest.TestCase)`.
- Here is the code for the test. You may add additional tests as well:

```python
import unittest

from ipp.exercises.labmodule09.SimpleKnnAlgorithm import SimpleKnnAlgorithm

class SimpleKnnAlgorithmTest(unittest.TestCase):
    """
    Simple set of tests for the SimpleKnnAlgorithm class.
    
    """
    
    def setUp(self):
        """Set up test fixtures before each test method."""
        self.knn = SimpleKnnAlgorithm(k = 3)
        
        # Sample training data: (features, target)
        self.regression_data = [
            ([1.0, 2.0], 10.0),
            ([2.0, 3.0], 15.0),
            ([3.0, 4.0], 20.0),
            ([4.0, 5.0], 25.0),
            ([5.0, 6.0], 30.0)
        ]
        
        # Classification data (using integer classes)
        self.classification_data = [
            ([1.0, 1.0], 0),
            ([1.5, 1.5], 0),
            ([2.0, 2.0], 0),
            ([8.0, 8.0], 1),
            ([8.5, 8.5], 1),
            ([9.0, 9.0], 1)
        ]
    
    def testDistanceCalculations(self):
        """Test both Euclidean and Manhattan distance calculations."""
        point1 = [0.0, 0.0]
        point2 = [3.0, 4.0]
        
        # Test Euclidean distance: sqrt(3^2 + 4^2) = 5.0
        euclidean = self.knn.calculateEuclideanDistance(point1, point2)
        self.assertAlmostEqual(euclidean, 5.0, places = 5)
        
        # Test Manhattan distance: |3| + |4| = 7.0
        manhattan = self.knn.calculateManhattanDistance(point1, point2)
        self.assertAlmostEqual(manhattan, 7.0, places = 5)
        
        # Test with identical points
        same_point = [1.0, 2.0]
        self.assertAlmostEqual(
            self.knn.calculateEuclideanDistance(same_point, same_point), 
            0.0, 
            places = 5
        )
        
        # Test with mismatched dimensions
        point3 = [1.0, 2.0, 3.0]
        distance = self.knn.calculateEuclideanDistance(point1, point3)
        self.assertEqual(distance, float('inf'))
    
    def testTrainingDataManagement(self):
        """Test setting and validating training data."""
        # Test setting valid training data
        self.knn.setTrainingData(self.regression_data)
        self.assertEqual(len(self.knn.trainingData), 5)
        self.assertEqual(self.knn.trainingData, self.regression_data)
        
        # Test with empty data
        empty_knn = SimpleKnnAlgorithm(k = 3)
        empty_knn.setTrainingData([])
        self.assertEqual(len(empty_knn.trainingData), 0)
        
        # Test with None
        none_knn = SimpleKnnAlgorithm(k = 3)
        none_knn.setTrainingData(None)
        self.assertEqual(len(none_knn.trainingData), 0)
    
    def testFindKNearest(self):
        """Test finding k nearest neighbors with both distance metrics."""
        self.knn.setTrainingData(self.regression_data)
        test_point = [2.5, 3.5]
        
        # Test with Euclidean distance
        neighbors_euclidean = self.knn.findKNearest(test_point, useEuclidean = True)
        self.assertEqual(len(neighbors_euclidean), 3)
        
        # Verify neighbors are sorted by distance
        for i in range(len(neighbors_euclidean) - 1):
            self.assertLessEqual(neighbors_euclidean[i][0], neighbors_euclidean[i + 1][0])
        
        # Test with Manhattan distance
        neighbors_manhattan = self.knn.findKNearest(test_point, useEuclidean = False)
        self.assertEqual(len(neighbors_manhattan), 3)
        
        # Test with no training data
        empty_knn = SimpleKnnAlgorithm(k = 3)
        neighbors_empty = empty_knn.findKNearest(test_point)
        self.assertEqual(neighbors_empty, [])
    
    def testRegressionPrediction(self):
        """Test regression prediction functionality."""
        self.knn.setTrainingData(self.regression_data)
        
        # Test prediction on a point between training points
        test_point = [2.5, 3.5]
        prediction = self.knn.predictRegression(test_point)
        self.assertIsInstance(prediction, float)
        self.assertGreater(prediction, 10.0)
        self.assertLess(prediction, 25.0)
        
        # Test on exact training point
        exact_point = [3.0, 4.0]
        exact_prediction = self.knn.predictRegression(exact_point)
        self.assertAlmostEqual(exact_prediction, 20.0, delta = 2.0)
        
        # Test with no training data
        empty_knn = SimpleKnnAlgorithm(k = 3)
        empty_prediction = empty_knn.predictRegression(test_point)
        self.assertIsNone(empty_prediction)
        
        # Test with Manhattan distance
        manhattan_prediction = self.knn.predictRegression(test_point, useEuclidean = False)
        self.assertIsNotNone(manhattan_prediction)
    
    def testClassificationPrediction(self):
        """Test classification prediction functionality."""
        self.knn.setTrainingData(self.classification_data)
        
        # Test point close to class 0
        test_point_class0 = [1.2, 1.2]
        prediction0 = self.knn.predictClassification(test_point_class0)
        self.assertEqual(prediction0, 0)
        
        # Test point close to class 1
        test_point_class1 = [8.2, 8.2]
        prediction1 = self.knn.predictClassification(test_point_class1)
        self.assertEqual(prediction1, 1)
        
        # Test with no training data
        empty_knn = SimpleKnnAlgorithm(k = 3)
        empty_prediction = empty_knn.predictClassification([1.0, 1.0])
        self.assertIsNone(empty_prediction)
        
        # Test with Manhattan distance
        manhattan_prediction = self.knn.predictClassification(
            test_point_class0, 
            useEuclidean = False
        )
        self.assertIn(manhattan_prediction, [0, 1])
    
    def testPredictionEvaluation(self):
        """Test prediction evaluation metrics."""
        # Test perfect prediction
        metrics_perfect = self.knn.evaluatePrediction(10.0, 10.0)
        self.assertEqual(metrics_perfect['error'], 0.0)
        self.assertEqual(metrics_perfect['absolute_error'], 0.0)
        self.assertEqual(metrics_perfect['squared_error'], 0.0)
        self.assertEqual(metrics_perfect['percent_error'], 0.0)
        
        # Test under-prediction
        metrics_under = self.knn.evaluatePrediction(10.0, 8.0)
        self.assertEqual(metrics_under['error'], 2.0)
        self.assertEqual(metrics_under['absolute_error'], 2.0)
        self.assertEqual(metrics_under['squared_error'], 4.0)
        self.assertAlmostEqual(metrics_under['percent_error'], 20.0, places = 1)
        
        # Test over-prediction
        metrics_over = self.knn.evaluatePrediction(10.0, 12.0)
        self.assertEqual(metrics_over['error'], -2.0)
        self.assertEqual(metrics_over['absolute_error'], 2.0)
        
        # Test with zero actual value and non-zero prediction
        metrics_zero = self.knn.evaluatePrediction(0.0, 5.0)
        self.assertEqual(metrics_zero['percent_error'], float('inf'))
        
        # Test with both zero
        metrics_both_zero = self.knn.evaluatePrediction(0.0, 0.0)
        self.assertEqual(metrics_both_zero['percent_error'], 0.0)


if __name__ == '__main__':
    unittest.main()
```

### Run Tests
- Testing using the `unittest` framework
  - In VS Code:
    - In the menu navigation on the left side, click on the beaker (mouseover the icons - hovering over the beaker will popup the name `Testing`)
    - All loaded tests will appear - click the `Refresh Testing` button to ensure they're all loaded.
      - NOTE: Tests must be in the top level `tests` path (within IPP_HOME) and each `unittest` module must begin with `test_` (e.g., `test_MyModule.py`)
    - Using the IDE's controls, execute the desired test(s). For this lab module, run the following:
      - `testDistanceCalculations`
      - `testTrainingDataManagement`
      - `testFindKNearest`
      - `testRegressionPrediction`
      - `testClassificationPrediction`
      - `testPredictionEvaluation`

- Testing the module directly
  - From within your IDE
    - Right click on your newly created module `test_SimpleKnnAlgorithm.py` and click `run tests`
    - You should see output similar to that discussed in class

  - From the command line
    - Open a terminal and cd to your `IPP_HOME` path
    - Start your virtual environment (if not already running)
    - Be sure your PYTHONPATH is set correctly
    - Run the module
      - ```python -m unittest ./tests/labmodule09/test_SimpleKnnAlgorithm.py```
    - You should see output similar to that discussed in class

### Sample output (yours may differ slightly)

```
Received test ids from temp file.
testClassificationPrediction (labmodule09.test_SimpleKnnAlgorithm.SimpleKnnAlgorithmTest)
Test classification prediction functionality. ... ok
testDistanceCalculations (labmodule09.test_SimpleKnnAlgorithm.SimpleKnnAlgorithmTest)
Test both Euclidean and Manhattan distance calculations. ... ok
testFindKNearest (labmodule09.test_SimpleKnnAlgorithm.SimpleKnnAlgorithmTest)
Test finding k nearest neighbors with both distance metrics. ... ok
testPredictionEvaluation (labmodule09.test_SimpleKnnAlgorithm.SimpleKnnAlgorithmTest)
Test prediction evaluation metrics. ... ok
testRegressionPrediction (labmodule09.test_SimpleKnnAlgorithm.SimpleKnnAlgorithmTest)
Test regression prediction functionality. ... KNN trained with 6 samples
KNN trained with 5 samples
KNN trained with 5 samples
ok
testTrainingDataManagement (labmodule09.test_SimpleKnnAlgorithm.SimpleKnnAlgorithmTest)
Test setting and validating training data. ... ok

----------------------------------------------------------------------
KNN trained with 5 samples
Warning: No training data provided
Warning: No training data provided
Ran 6 tests in 0.001s

OK
Finished running tests!
```
